File: LinkedList.java
import java.util.NoSuchElementException;

public class LinkedList {
    private class Node {
        private int value;
        private Node next;

        public Node(int value) {
            this.value = value;
        }
    }

    private Node first;
    private Node last;

    public void addLast(int item) {
        var node = new Node(item);

        if (isEmpty())
        first = last = node;
        else {
            node.next = first;
            first = node;
        }
    }

    private boolean isEmpty() {
        return first == null;
    }

    public int indexOf( int item ) {
        int index = 0;
        var current = first;
        // boolean currentIsNotNull = current != null;
        // TODO: figure out why can't use both boolean variables
        // boolean currentValueIsEqualToItem = current.value == item;

        while ( current != null ) {
            if ( current.value == item ) return index;
            // setting current to the next node
            current = current.next;
            // incrementing the index
            index++;
        }
        // if we can't find node with this value
        return -1;
    }

    public boolean contains(int item) {
        return indexOf(item) != -1;
    }

    public void removeFirst() {
        // [10 -> 20 -> 30]
        if (isEmpty())
            throw new NoSuchElementException();
        

        if (first == last) {
            first = last = null;
            return;
        }

        var second = first.next;
        first.next = null;
        first = second;
    }

    public void removeLast() {
        // [10 -> 20 -> 30]
        var previous = getPrevious(last);
        last = previous;
        last.next = null;
    }

    private Node getPrevious(Node node) {
        var current = first;
        while (current != null) {
            if (current.next == node) return current;
            current = current.next;
        }
        return null;
    }
}

File: linkedlist.py
class Node:
    def __init__ (self, data):
        self.data=data
        self.next=None

class LinkedList:
    def __init__ (self):
        self.head=None
    
    def reverse(self):
        prev=None
        current=self.head
        while(current is not None):
            next=current.next
            current.next=prev
            prev=current
            current=next
        self.head=prev

    def push(self, newData):
        newNode=Node(newData)
        newNode.next=self.head
        self.head=newNode

    def printList(self):
        temp=self.head
        while(temp):
            print(temp.data)
            temp=temp.next

list=LinkedList()
list.push(34)
list.push(545)
list.push(3)
list.push(9)
list.push(45)

list.printList()
list.reverse()
list.printList()